sale_picking_state/models/stock_picking.py

```diff
@@ -0,0 +1,27 @@
+# -*- coding: utf-8 -*-
+# Copyright 2018 Akretion (http://www.akretion.com).
+# @author Pierrick BRUN <pierrick.brun@akretion.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import models, api
+
+
+class StockPicking(models.Model):
+    _inherit = 'stock.picking'
+
+    @api.multi
+    def write(self, vals):
+        result = super(StockPicking, self).write(vals)
+        if 'date_done' in vals:
+            for picking in self:
```

Comment:
for sale in self.mapped('sale_id'):
    states = sale.mapped('picking_ids.state')

if states ....